### PR TITLE
bugfix: symbol would be encoded to 0x3f when using proxy-rewrite directive

### LINKED ISSUE #1801: request help: how to rewrite query parameters using proxy-rewrite
### Issue description

'?' symbol would be encoded to 0x3f when using proxy-rewrite directive. 

https://github.com/apache/incubator-apisix/blob/794b1eb65e01e0e15d3dcd9cf1f0d74166e9b246/apisix/plugins/proxy-rewrite.lua#L156

Is there any way to rewrite url to append some more query parameter to the original path?

It would be highly appreciated if rewriting query parameter could be supported like in nginx: https://serverfault.com/questions/311487/nginx-rewrite-append-a-parameter-at-the-end-of-an-url

### Environment

* apisix version (cmd: 1.3):
* OS: Linux

### PR DESCRIPTION
**Pull request by wrl96**
### Summary

bugfix: symbol would be encoded to 0x3f when using proxy-rewrite directive

### Full changelog

* fix #1801 
* add a test case

### Issues resolved

Fix #1801

### CODE REVIEW

File: apisix/plugins/proxy-rewrite.lua [hunk 1]
```diff
@@ -153,10 +154,20 @@ function _M.rewrite(conf, ctx)
         end
     end
 
-    upstream_uri = core.utils.uri_safe_encode(upstream_uri)
+    local index, _ = find_str(upstream_uri, "?")
+    if index then
+        upstream_uri = core.utils.uri_safe_encode(sub_str(upstream_uri, 1, index-1)) ..
```
Comment (membphis) [inline apisix/plugins/proxy-rewrite.lua:159]: missing `?`

Comment (wrl96) [reply apisix/plugins/proxy-rewrite.lua:159]: sub_str(upstream_uri, index) start with `?`

Comment (wrl96) [reply apisix/plugins/proxy-rewrite.lua:159]: > missing `?`

sub_str(upstream_uri, index) start with `?`

File: apisix/plugins/proxy-rewrite.lua [hunk 2]
```diff
@@ -153,10 +154,20 @@ function _M.rewrite(conf, ctx)
         end
     end
 
-    upstream_uri = core.utils.uri_safe_encode(upstream_uri)
+    local index, _ = find_str(upstream_uri, "?")
```
Comment (spacewander) [inline apisix/plugins/proxy-rewrite.lua:157]: Should use plain text style `string.find`.

File: apisix/plugins/proxy-rewrite.lua (hunk 1 above)
Comment (membphis) [reply apisix/plugins/proxy-rewrite.lua:159]: I think your way is correct. Ignore what I said.

File: apisix/plugins/proxy-rewrite.lua (hunk 2 above)
Comment (wrl96) [reply apisix/plugins/proxy-rewrite.lua]: > Should use plain text style `string.find`.

Edited.

File: apisix/plugins/proxy-rewrite.lua [hunk 3]
```diff
@@ -153,10 +154,20 @@ function _M.rewrite(conf, ctx)
         end
     end
 
-    upstream_uri = core.utils.uri_safe_encode(upstream_uri)
+    local index = find_str(upstream_uri, "?", 1, true)
+    if index then
+        upstream_uri = core.utils.uri_safe_encode(sub_str(upstream_uri, 1, index-1)) ..
+        sub_str(upstream_uri, index)
```
Comment (membphis) [inline apisix/plugins/proxy-rewrite.lua:160]: bad indentation

Comment (wrl96) [reply apisix/plugins/proxy-rewrite.lua]: > bad indentation

Fixed.

File: apisix/plugins/proxy-rewrite.lua [hunk 4]
```diff
@@ -153,10 +154,20 @@ function _M.rewrite(conf, ctx)
...
+        upstream_uri = core.utils.uri_safe_encode(sub_str(upstream_uri, 1, index-1)) ..
+                       sub_str(upstream_uri, index)
+    else
+        upstream_uri = core.utils.uri_safe_encode(upstream_uri)
+    end
 
     if ctx.var.is_args == "?" then
-        ctx.var.upstream_uri = upstream_uri .. "?" .. (ctx.var.args or "")
+        if index then
+            ctx.var.upstream_uri = upstream_uri .. "&" .. (ctx.var.args or "")
```
Comment (moonming) [inline apisix/plugins/proxy-rewrite.lua:167]: I'm wondering if we need to keep the request args?

Comment (wrl96) [reply apisix/plugins/proxy-rewrite.lua:167]: > I'm wondering do we need to keep the request args?

I think we need to keep it.
If someone needn't it, he can remove args by regex_uri.

**Comment by membphis** [comment]
@wrl96 https://travis-ci.org/github/apache/incubator-apisix/jobs/706394034#L991

you need to rebase your branch, that is the bug of Travis CI. we can not merge this PR if the CI is failed.

**Comment by wrl96** [comment]
> @wrl96 https://travis-ci.org/github/apache/incubator-apisix/jobs/706394034#L991
> 
> you need to rebase your branch, that is the bug of Travis CI. we can not merge this PR if the CI is failed.

Fixed

**Comment by membphis** [comment]
@wrl96 many thx, merged

BTW, you should use `git rebase`

**Comment by wrl96** [comment]
> @wrl96 many thx, merged
> 
> BTW, you should use `git rebase`

ok, I used ` Rebase my Github fork` in idea.

**Pull request merged**